File: 567-permutation-in-string/permutation-in-string.java
class Solution {
    public boolean checkInclusion(String s1, String s2) {
        int[] f1 = new int[26];
        int[] f2 = new int[26];
        if(s1.length()>s2.length()) return false;
        for(int i=0;i<s1.length();i++){
            f1[s1.charAt(i)-'a']++;
            f2[s2.charAt(i)-'a']++;
        }
        int count=0;
        for(int i=0;i<26;i++){
            if(f1[i]==f2[i]) count++;
        }
        if(count==26) return true;
        for(int i = s1.length();i<s2.length();i++)
        {
            int before = i-s1.length();
            char be = s2.charAt(before);
            int now = i;
            char no = s2.charAt(now);
            //1.if it was a match before since we are moving 1 step 
            //count--
            if(f1[be-'a']==f2[be-'a']) count--;
            //2. If current added element freq before updating is match
            // now if won't be match now
            if(f1[no-'a']==f2[no-'a']) count--;

            f2[no-'a']++;//adding element
            f2[be-'a']--;//removing element

            if(f1[no-'a']==f2[no-'a']) count++;
            if(f1[be-'a']==f2[be-'a']) count++;

            if(count==26) return true;

        }
        return false;
        
    }
}

File: 567-permutation-in-string/permutation-in-string.cpp
class Solution {
public:
    bool checkInclusion(string s1, string s2) {
        vector<int> a(26,0);
        vector<int> b(26,0);
        int count =0;
        int k = s1.size();

        if(k>s2.size()) return false;
        int j=0;
        for(char c: s1){
            int i = c-'a';
               a[i]+=1;
        }
        for(j=0;j<s1.size();j++){
             char c = s2[j];
             int i = c-'a';
             b[i]++;
        }
        int i=0;
        while(i<26){
            if(a[i]==b[i]) count++;
            i++;
        }

         while(j < s2.size()){
            if(count==26) return true;
            int current = s2[j]-'a';
            int before = s2[j-k]-'a';
             // moving window by 1 
            if(b[before]==a[before]) count--; //if previous was a match now it won't be match
            b[before]--;
            if(b[before]==a[before]) count++; // after removal its a match so incerement count by 1

            if(b[current]==a[current]) count--; //if before frequency update it's a match now it won't be a match
            b[current]++;
            if(b[current]==a[current]) count++;// after freq update i'ts a match now
            j++;
        }
        if(count==26)return true;
        return false;
    }
};